cdc/cdc.cc

```diff
@@ -194,4 +195,194 @@ future<> setup(service::storage_proxy& proxy, schema_ptr s) {
     });
 }
 
+class transformer final {
+public:
+    struct map_cmp {
+        bool operator()(const std::pair<net::inet_address, unsigned int>& a,
+                        const std::pair<net::inet_address, unsigned int>& b) const {
+            if (a.first.is_ipv4()) {
+                if (b.first.is_ipv4()) {
+                    if (a.first.as_ipv4_address().ip > b.first.as_ipv4_address().ip) {
+                        return false;
+                    }
+                    return a.second < b.second;
+                } else {
+                    return true;
+                }
+            } else {
+                if (b.first.is_ipv4()) {
+                    return false;
+                } else {
+                    if (a.first.as_ipv6_address().ip > b.first.as_ipv6_address().ip) {
+                        return false;
+                    }
+                    return a.second < b.second;
+                }
+            }
+        }
+    };
+    using streams_type = std::map<std::pair<net::inet_address, unsigned int>, utils::UUID, map_cmp>;
+private:
+    schema_ptr _schema;
+    schema_ptr _log_schema;
+    utils::UUID _time;
+    bytes _decomposed_time;
+    std::reference_wrapper<locator::token_metadata> _token_metadata;
+    std::reference_wrapper<const locator::abstract_replication_strategy> _replication_strategy;
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    unsigned _ignore_msb_bits;
+    streams_type _streams;
+
+    clustering_key set_pk_columns(const partition_key& pk, int batch_no, mutation& m) const {
+        const auto log_ck = clustering_key::from_exploded(
+                *m.schema(), { _decomposed_time, int32_type->decompose(batch_no) });
+        auto pk_value = pk.explode(*_schema);
+        size_t pos = 0;
+        for (const auto& column : _schema->partition_key_columns()) {
+            assert (pos < pk_value.size());
+            auto cdef = m.schema()->get_column_definition(to_bytes("_" + column.name()));
+            auto value = atomic_cell::make_live(*column.type,
+                                                _time.timestamp(),
+                                                bytes_view(pk_value[pos]));
+            m.set_cell(log_ck, *cdef, std::move(value));
+            ++pos;
+        }
+        return log_ck;
+    }
+    partition_key stream_id(const schema& s, const net::inet_address& ip, unsigned int shard_id) const {
```

Comment:
suggestion: don't take the `schema` argument, since it's always equal to `*_log_schema`.